### DISCUSSION TITLE
Enable Writing of Custom Layouts

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#17426] [Q&A / Support]
How do I enable writing / saving of custom layouts in v11?

I see this code: https://github.com/pimcore/admin-ui-classic-bundle/blob/e9a696d89315253d5dbcf34b83ec36070a6fabac/src/Controller/Admin/IndexController.php#L275

but am not sure of what env var, etc would enable it.

i.e. Is there an equivalent of `PIMCORE_CLASS_DEFINITION_WRITABLE` for custom layouts?

**Comment by kingjia90** [comment]
You might be looking for store location [symfony-config](https://github.com/pimcore/pimcore/blob/e6874414821637d385df6c08606db8804f610e8b/doc/21_Deployment/03_Configuration_Environments.md?plain=1#L91) which saves the configuration in a yaml file instead of the db table `settings_store`

**Reply by cancan101** [reply]
Is it possible to set this using env vars? If so what would the env var be:
```
PIMCORE_WRITE_TARGET_CUSTOMLAYOUT
PIMCORE_WRITE_TARGET_CUSTOM_LAYOUT
PIMCORE_WRITE_TARGET_CUSTOM_LAYOUTS
PIMCORE_WRITE_TARGET_OBJECT_CUSTOMLAYOUT
PIMCORE_WRITE_TARGET_DATAOBJECT_CUSTOM_LAYOUT
```
etc? From https://github.com/pimcore/pimcore/blob/e6874414821637d385df6c08606db8804f610e8b/doc/21_Deployment/03_Configuration_Environments.md?plain=1#L38 hard to tell what the name would be.

**Reply by kingjia90** [reply]
Not out of the box from the core, but you can set an own custom env variable like

//.env
CUSTOMLAYOUT=symfony-config

//someconfig.yaml like example above
...
object_custom_layouts:
write_target:
   type:'%env(string:CUSTOMLAYOUT)%' 
...
and should work

**Reply by cancan101** [reply]
Would it make sense that when `PIMCORE_CLASS_DEFINITION_WRITABLE=true`, custom layouts can also be modified? It would seem odd that we want to allow live modification of the class itself but not the layouts.

**Reply by kingjia90** [reply]
Iirc it is intended to be like this because devs may want to disable the possibility to change class definitions in production but still want to allow to create/edit custom layouts, so by keeping it independent from the env variable, it has more flexibility in terms of possible "combinations"

**Reply by cancan101** [reply]
Do any of the of the env var like `PIMCORE_WRITE_TARGET_DATA_HUB` still work?  I know they used to, see: https://github.com/orgs/pimcore/discussions/10940#discussioncomment-1736286.

Looks like no longer: https://pimcore.com/docs/platform/2023.3/Pimcore/Installation_and_Upgrade/Updating_Pimcore/Preparing_for_V11/#configuration-adaptions

**Reply by cancan101** [reply]
Is there anyway with the env processor to have defaults? Something like this where the fallback is to `symfony-config`?
```
type:'%env(string:CUSTOMLAYOUT:symfony-config)%'
````

**Reply by kingjia90** [reply]
it should be possible to have [defaults](https://symfony.com/blog/new-in-symfony-4-3-default-and-trim-env-var-processors#default-env-var-processor), but it sounds like going into a XY problem, what's the initial purpose for this?

**Reply by cancan101** [reply]
The purpose of all this is to have different settings on our qa / staging site from production and to use env car rather than files to control that behavior (ie 12 factor). For example, on the staging site we want to allow changing of settings by users but on the production site we want to lock things down. 

The 12-factor approach is more natively supported before the env var setting was removed. The worker on right now is just to make sure that the env var does set  on all sites and not to rely on there being a default in the config file

**Reply by kingjia90** [reply]
ahh, i have never heard of this "12 factor" before, i see now, that's really interesting from a quick glance, but makes me wonder how easily Symfony itself can natively adhere to this manifesto as it is more revolving around to the `Configuration Environment` (eg. dev/prod/test) and all its override and fallback. 

So i believe the link above with the symfony default and env processor can be achieved but looks gonna be a massive and long re-invention, hopefully somebody from the community that had more experience on this can advise.

**Reply by cancan101** [reply]
As of now the env approach as listed above does not work due to this issue: https://github.com/pimcore/pimcore/issues/15980.
